### PR TITLE
azurerm_kubernetes_cluster - support for the `only_critical_addons_enabled` property

### LINKED ISSUE #9183: Support for AKS node_taints in default_node_pool again
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I understand the background of #8982 , then node_taints is no longer possible to configure from v2.35. But, AKS API allow exception for CriticalAddonsOnly taint on system nodepools (or all) from the latest update.

https://github.com/Azure/AKS/issues/1833

This option is very useful and critical, so could you please consider support it again?

### New or Affected Resource(s)

* azurerm_kubernetes_cluster

### Potential Terraform Configuration

```hcl
  default_node_pool {
    node_taints = ["CriticalAddonsOnly=true:PreferNoSchedule"]
  }
```

### References

* #8982

### PR DESCRIPTION
**Pull request by favoretti**
Fixes #9183

### CODE REVIEW

File: azurerm/internal/services/containers/kubernetes_nodepool.go [hunk 1]
```diff
@@ -372,6 +383,11 @@ func FlattenDefaultNodePool(input *[]containerservice.ManagedClusterAgentPoolPro
 		}
 	}
 
+	criticalAddonsEnabled := false
+	if agentPool.NodeTaints != nil {
+		criticalAddonsEnabled = true
```
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_nodepool.go:388]: this isn't strictly true that this is enabled by virtue of their being a taint (incase someone cycles the default node pool outside of Terraform, which we don't support at this point, but is a longer term thing to solve) - instead we can guarantee this by checking the node taints contain this value:

```suggestion
		for _, taint := range *agentPool.NodeTaints {
			if strings.EqualsFold(taint, "CriticalAddonsOnly=true:PreferNoSchedule") {
				criticalAddonsEnabled = true
			}
		}
```

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 2]
```diff
@@ -223,6 +223,8 @@ A `default_node_pool` block supports the following:
 
 -> **NOTE:** This requires that the `type` is set to `VirtualMachineScaleSets` and that `load_balancer_sku` is set to `Standard`.
 
+* `only_critical_addons_enabled` - (Optional) Enabling this option will taint default node pool with `CriticalAddonsOnly=true:PreferNoSchedule` taint. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/kubernetes_cluster.html.markdown:226]: _minor_ but could we sort this alphabetically to match the rest of this file (move this a few lines down)

File: azurerm/internal/services/containers/kubernetes_nodepool.go [hunk 3]
```diff
@@ -153,6 +153,11 @@ func SchemaDefaultNodePool() *schema.Schema {
 					ForceNew:     true,
 					ValidateFunc: computeValidate.ProximityPlacementGroupID,
 				},
+				"only_critical_addons_enabled": {
```
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_nodepool.go:156]: could we add an acceptance test creating a cluster where this is enabled for the default node pool?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @favoretti 

Thanks for this PR / pushing those changes - taking a look through if we can fix up the minor comments then this otherwise LGTM 👍 

Thanks!

File: azurerm/internal/services/containers/kubernetes_nodepool.go (hunk 1 above)
Comment (favoretti) [reply azurerm/internal/services/containers/kubernetes_nodepool.go:388]: I thought that I wanted to get out of this too easy :) Good call, thank you.

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 2 above)
Comment (favoretti) [reply website/docs/r/kubernetes_cluster.html.markdown:226]: Sigh, I actually did sort it alphabetically to only then correct the fact that it wasn't `critical_addons_enabled`, but `only_critcal_addons_enabled` :D

**Comment by favoretti** [comment]
```
$ TF_ACC=1 go test -v ./azurerm/internal/services/containers -timeout=1000m -run=TestAccKubernetesCluster_criticalAddonsTaint
=== RUN   TestAccKubernetesCluster_criticalAddonsTaint
=== PAUSE TestAccKubernetesCluster_criticalAddonsTaint
=== CONT  TestAccKubernetesCluster_criticalAddonsTaint
--- PASS: TestAccKubernetesCluster_criticalAddonsTaint (691.63s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers	692.807s
```

Resource JSON - Microsoft Azure 2021-01-27 10-38-14

File: azurerm/internal/services/containers/kubernetes_nodepool.go (hunk 3 above)
Comment (favoretti) [reply azurerm/internal/services/containers/kubernetes_nodepool.go:157]: Certainly, added, results in the comment.

**Review by tombuildsstuff — APPROVED** [review]
LGTM - thanks for pushing those changes @favoretti

**Comment by t3mi** [comment]
@tombuildsstuff sorry for chiming in but these changes aren't correct as API allows multiple soft taints + strict hard taint.

**Comment by tombuildsstuff** [comment]
@t3mi can you give a little more context? The default node pool and additional node pools have different constraints for these fields - so this is only affecting the Default Node Pool?

**Comment by t3mi** [comment]
I'm talking about default node pool as I haven't checked constraints for ordinary ones.
Relevant issues - [AKS#1833](https://github.com/Azure/AKS/issues/1833), [AKS#1484](https://github.com/Azure/AKS/issues/1484)
Shortly speaking, API allows settings n number of soft taints (the ones with PreferNoSchedule effect) while only one hard taint is allowed - `CriticalAddonsOnly=true:NoSchedule`. Also setting hardcoded taint with `PreferNoShedule` doesn't make sense as it won't prevent user pods to be scheduled on the default node pool as was described in primary use case for this.

**Comment by favoretti** [comment]
@t3mi If I remove `PreferNoSchedule`, from the taint, i.e. make it a hard taint - would this be acceptable for now? We can talk about other soft taints in a separate PR? :)

**Comment by t3mi** [comment]
@favoretti I'm not a decision maker here, just shared my concerns. From my point it doesn't make sense in introducing separate parameter just to remove it later for soft taints as its not needed. Soft taints are the one with `PreferNoSchedule`, for hard taint effect is `NoSchedule`.  Hard taint could be only with key value `CriticalAddonsOnly`, for soft taint there is no such restriction. Both of them could be combined and placed on a default node pool.

Validation from the API just for reference:
```Placing custom taints on system pool is not supported(except 'CriticalAddonsOnly' taint or taint effect is 'PreferNoSchedule'). Please refer to https://aka.ms/aks/system-taints for detail"```

**Comment by favoretti** [comment]
@t3mi Ok, I flipped this option to a hard taint for now, to allow for this to make it to the next release and for people to keep their clusters clean, let's talk about reintroducing general soft taints on default node pool in a separate change?

**Comment by favoretti** [comment]
```
$ TF_ACC=1 go test -v ./azurerm/internal/services/containers -timeout=1000m -run=TestAccKubernetesCluster_criticalAddonsTaint
=== RUN   TestAccKubernetesCluster_criticalAddonsTaint
=== PAUSE TestAccKubernetesCluster_criticalAddonsTaint
=== CONT  TestAccKubernetesCluster_criticalAddonsTaint
--- PASS: TestAccKubernetesCluster_criticalAddonsTaint (649.52s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers	650.759s
```

Acceptance test still passes after the last changes.

**Comment by dhirschfeld** [comment]
It seems this missed the boat for 2.46 :(

Since the CI is green, is there anything preventing this PR being merged so it can make it in the next release?

**Comment by luddskunk** [comment]
I'm also wondering the same. I was tracking it yesterday and it was there as green.

Could you please elaborate why it was moved to v.2.47.0 @tombuildsstuff ? Thanks!

**Comment by favoretti** [comment]
Mainly due to $reasons and the fact that acceptance tests take over 12 hours. I hope Tom will merge it early next week and it ships Thursday.

**Review by katbyte — APPROVED** [review]
Thanks @favoretti - LGTM 👍

**Comment by Cyanopus** [comment]
@favoretti  We tried internally with this version and the error persists.

"Error: expanding `default_node_pool`: The AKS API has removed support for tainting all nodes in the default node pool and it is no longer possible to configure this. To taint a node pool, create a separate one"

**Comment by favoretti** [comment]
@Cyanopus `node_labels` and `node_taints` require nodepool to be recreated, so `only_critical_addons_enabled` is no different. Azure API doesn't support patching those, so unfortunately there's nothing I can do about a reroll, I'm sorry.

**Pull request merged**